### ISSUE TITLE
Porting PTPd to Win32 compatible OS

### ISSUE THREAD

**Issue by eduardodsp** [ptpd/ptpd#73]
Hi, I've been working with an Win32 compatible RTOS on an industrial application and I am wondering if it would be possible to port ptpd's current release to such OS. 

If it is possible, I wonder if anyone could give me some advice, some "where to start" kind of thing.

best regards!

**Comment by nallaravi** [comment]
Hi, 
Please share if you found out some information.
Thanks!

**Comment by j123b567** [comment]
@eduardodsp what RTOS?

Several years ago, we try to port PTPd to these Win32 API "compatible" systems.
 
 - PharLap RTOS - it lacks essential features like UDP multicast and time manipulation so we abandon the project
 - IntervalZero RTX - it lacks essential features to manipulate time so it forces us to reimplement all timescale manipulating functions using `QueryPerformanceCounter` and every tasks need to use our functions. It was also real pain to work with RTX because of Win32 like API. RTX functions shared similar name but behaves differently resulting in lot of "green screans of death" during development. Final precision of comparing hardware PPS signals was acceptable - in terms of microseconds. It is possible, that it has now these features or even native PTP support but I'm no more interested.
 - Windows 7 - it laks essential features to precisely adjust time so we achieve precision about 1ms which is far behind other *nix implementations.  Maybee, I can find some code for this port. Windows 10 has finally essential functions `SetSystemTimeAdjustmentPrecise` and `GetSystemTimePreciseAsFileTime`. `FILETIME` is still 2 orders worse then `timespec` in *nix but still better then nothing. I never tried these features anyway.

**Comment by eduardodsp** [comment]
The RTOS in question is On-Time's RTOS-32. It's worth mentioning that with RTOS-32 you don't have to use it's Win32 emulation layer, it's native API library is far more powerful.

**Comment by nallaravi** [comment]
Thanks @j123b567. Could you share the build instructions for Windows 10?

**Comment by j123b567** [comment]
No. I don't have Windows 10 version. Windows 10 introduces new API to finally support better time synchronization.
I have only very outdated (>8 years) PTPd version with many bugs for Windows 7 which works very poorly - it is better to use NTP then this.

But you can have a look and try to port current version yourself with hints from this version https://github.com/j123b567/ptpd-old-work/tree/win32 and new Win API (`SetSystemTimeAdjustmentPrecise` and `GetSystemTimePreciseAsFileTime`)

**Comment by nt2ds** [comment]
Hi, I have almost finished an entirely new PTPv2 daemon from scratch. All that is remaining is the Servo mechanism and I will upload the source code. I might finish it by the end of the week. It is going to be released on the end of the month though no matter what. It  is part of a bigger project, AES67 audio transmitter. My daemon currently features multithreading via IOCP so you can integrate it into another app easily without the hassle of trying to run it on a separate thread for non blocking functionality

**Comment by nt2ds** [comment]
Hi guys. Its ready.
https://github.com/nt2ds/Win32_PTP
